lib/net.js

```diff
@@ -425,12 +421,14 @@ Socket.prototype._destroy = function(exception, cb) {
   var self = this;
 
   function fireErrorCallbacks() {
+    var hadSeenError = self._errorEmitted;
+    self._errorEmitted = self._errorEmitted || Boolean(exception);
```

Comment:
This moves the `_errorEmitted` flag functionality down to the `net.Socket` class, since only it and `tls.TLSSocket` are making use of it, and are able to set it themselves apart from the parent class. This is part of a larger effort to reduce the number of places that stream subclasses reach into parent private state.